### PR TITLE
Sprint Snapshots

### PR DESCRIPTION
**Pull request by jakobw**
@JeroenDeDauw ping!

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
Will try to review this tomorrow. Might not work, in case I'll do it on Friday.

File: app/models/SprintSnapshot.php [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+class SprintSnapshot extends Eloquent {
+	protected $fillable =  ['data', 'sprint_id'];
+
+	public function sprint()
```
Comment (JeroenDeDauw) [inline app/models/SprintSnapshot.php]: `getSprint` or something like `getParentSprint`? Also, what does this return? A `Sprint` or some integer id?

File: app/models/Sprint.php [hunk 2]
```diff
@@ -14,6 +15,11 @@ public function project()
 		return $this->belongsTo('Project');
 	}
 
+	public function sprintSnapshots()
```
Comment (JeroenDeDauw) [inline app/models/Sprint.php]: `getSprintSnapshots`

This returns an array of ... ?

File: app/models/Sprint.php [hunk 3]
```diff
@@ -92,4 +98,36 @@ private function createPhabricatorProject()
 
 		return true;
 	}
+
+	/**
+	 * @return bool Returns whether the snapshot was successfully created
+	 */
```
Comment (JeroenDeDauw) [inline app/models/Sprint.php]: Do you need to do this return for some reason? It's a classical anti-pattern violating the [CQS principle](http://en.wikipedia.org/wiki/Command%E2%80%93query_separation)

File: app/Phragile/BurndownChart.php [hunk 4]
```diff
@@ -9,13 +9,13 @@ class BurndownChart {
 	private $pointsClosedPerDay = null;
 	private $sprint = null;
 	private $tasks = null;
-	private $phabricator = null;
+	private $transactions = null;
 
-	public function __construct(\Sprint $sprint, TaskList $tasks, PhabricatorAPI $phabricator)
+	public function __construct(\Sprint $sprint, TaskList $tasks, array $transactions)
```
Comment (JeroenDeDauw) [inline app/Phragile/BurndownChart.php]: \o/

now It's not entirely obvious what type the parameter is now though. I'm assuming it's an array of `Transaction`, though it's good to document this.

File: app/Phragile/BurndownChart.php [hunk 5]
```diff
@@ -38,26 +38,12 @@ private function calculatePointsClosedPerDay()
...
-				return $task['closed'];
-			})
-		);
-	}
-
 	private function getClosedTaskTimes()
 	{
-		$taskTransactions = $this->phabricator->getTaskTransactions($this->getClosedTaskIDs());
-
 		return array_map(function($transactions)
```
Comment (JeroenDeDauw) [inline app/Phragile/BurndownChart.php]: Should the parameter of the anon function not be called `$transaction`? Or is the transcations field an array of arrays?

File: app/models/Sprint.php (hunk 2 above)
Comment (jakobw) [inline app/models/Sprint.php]: It's a laravel thing which lets me access the SprintSnapshots of a Sprint through a dynamic property http://laravel.com/docs/4.2/eloquent#dynamic-properties

File: app/Phragile/BurndownChart.php (hunk 5 above)
Comment (jakobw) [inline app/Phragile/BurndownChart.php]: It's an array of arrays containing the transactions for each task.

File: app/models/Sprint.php (hunk 2 above)
Comment (JeroenDeDauw) [inline app/models/Sprint.php]: So after quickly looking at this, I guess it returns a `Collection` which contains `SprintSnapshot` instances. Perhaps this is obvious if you know the framework. Still useful for people that don't.

**Pull request merged**